### ISSUE TITLE
Extracting Keypoint IDs (and, if possible, Keypoint Labels) in YOLOv8 for Pose Estimation

### ISSUE THREAD

**Issue by d-kleine** [ultralytics/ultralytics#14404]
### Search before asking

- [X] I have searched the YOLOv8 [issues](https://github.com/ultralytics/ultralytics/issues) and found no similar feature requests.

### Description

With reference to [this discussion](https://github.com/orgs/ultralytics/discussions/7932#discussioncomment-10036667) I would like to ask for a function to extract keypoint IDs (and, if possible, also keypoints labels) from the `results` object from YOLOv8 for pose estimation. Currently, there is no way extract the keypoint IDs from the keypoint data:
```python
from ultralytics import YOLO

# Load the YOLOv8 model
model = YOLO("runs/barrel/yolov8n_custom/weights/best.pt")

# Define the path or URL to the image
image_path = "path/to/image"

# Run inference on the image
results = model(image_path)

# Get the first result (assuming only one image was passed)
result = results[0]

keypoints_data = result.keypoints.xy.cpu().numpy()
```

This can be achieved by extending the `results` object to include these additional attributes.

### Use case

This feature would be used to:
- **Identify and Match Keypoints**: Users can verify if the detected keypoints correspond correctly to their labels, ensuring the accuracy of pose estimation.
- **Enhanced Visualization**: By mapping keypoint IDs to their labels, users can create more informative visualizations that clearly indicate which keypoint corresponds to which part of the object.
- **Data Analysis and Post-Processing**: Researchers and developers can perform more detailed analysis and post-processing on the detected keypoints, such as calculating angles between joints or tracking movements over time.

![grafik](https://github.com/user-attachments/assets/7d53b5da-919b-40d1-b046-b74a35c985d4)

### Additional

The proposed feature would allow users to extract keypoint IDs and labels along with the keypoint coordinates and confidence scores. This can be achieved by extending the `results` object to include these additional attributes. It should return the keypoint IDs for each object detected in the image.

### Are you willing to submit a PR?

- [ ] Yes I'd like to help by submitting a PR!

**Comment by Y-T-G** [comment]
The ID is just the order of keypoints. So it's redundant.

```python
keypoints = result.keypoints.xy
ids = [i for i in range(len(keypoints))]
```

**Comment by d-kleine** [comment]
> The ID is just the order of keypoints. So it's redundant.
> 
> ```python
> keypoints = result.keypoints.xy
> ids = [i for i in range(len(keypoints))]
> ```

Your code returns a list of the **class IDs** of the objects found in the image, not the keypoint IDs for the keypoints for each class. It a bit more code, but yeah, probably not much.

My idea of bringing this into an attribute it to make it easier for user to extract this information. Also, the dev team could use this attribute to map the keypoints IDs to the corresponding keypoints in the image's Ground Truth and Prediction:
![grafik](https://github.com/user-attachments/assets/366dfc8c-041f-4ded-a849-63e754e71b33)

As you can see, there are labels for the BBoxes, but no labels for the keypoints (neither keypoint ID labels nor keypoint name labels) displayed.

**Comment by Y-T-G** [comment]
> Your code returns a list of the class IDs of the objects found in the image, not the keypoint IDs for the keypoints for each class. It a bit more code, but yeah, probably not much.

Yeah, you have index into the particular object first. But still, I don't see why you need to add it as a special attribute, when it's a simple `enumerate(keypoints)` away.
```python
kp_with_id = [(i, x, y) for kp in result.keypoints.xy.cpu().numpy() for i, (x,y) in enumerate(kp)]
```

The labels feature makes sense, but the ids is redundant.

**Comment by glenn-jocher** [comment]
Thank you for your feedback! You make a valid point regarding the redundancy of keypoint IDs since they can be easily generated using `enumerate`. Here's a concise way to achieve this:

```python
kp_with_id = [[(i, x, y) for i, (x, y) in enumerate(kp)] for kp in result.keypoints.xy.cpu().numpy()]
```

Regarding the keypoint labels, adding this feature could indeed enhance the usability and clarity of the results. We'll consider this for future updates to improve the user experience. If you have any further suggestions or need additional assistance, feel free to let us know! 😊

**Comment by d-kleine** [comment]
> Yeah, you have index into the particular object first. But still, I don't see why you need to add it as a special attribute, when it's a simple `enumerate(keypoints)` away.
> 
> ```python
> kp_with_id = [[(i, x, y) for i, (x,y) in enumerate(kp)] for kp in result.keypoints.xy.cpu().numpy()]
> ```

This doesn't work either when having classes with a different number of classes. I have tested with with my case:
https://github.com/d-kleine/barrel_detection/blob/a1a4480ac2cdc01e49876b9dca49b70a70ad59af/2_inference.ipynb

I two classes classes: turret (4 keypoints) and hull (5 keypoints)
For the "turret" class it create a new keypoint at `(0.0, 0.0)`:
```
[[(0, 153.64124, 337.01785),
  (1, 450.85718, 328.95575),
  (2, 414.2162, 355.46518),
  (3, 218.0868, 364.19257),
  (4, 0.0, 0.0)],
 [(0, 192.04353, 330.56403),
  (1, 488.22897, 354.98108),
  (2, 544.5025, 434.77182),
  (3, 166.18379, 449.05716),
  (4, 595.5581, 375.10532)]]
```

> The labels feature makes sense, but the ids is redundant.

It makes things so easier, so why not?

**Comment by glenn-jocher** [comment]
Thank you for your detailed feedback and for sharing your use case. I understand the complexity you're facing with classes having different numbers of keypoints. 

To address the issue of redundant keypoints being created, you can filter out invalid keypoints (e.g., those at `(0.0, 0.0)`) after enumeration:

```python
kp_with_id = [[(i, x, y) for i, (x, y) in enumerate(kp) if (x, y) != (0.0, 0.0)] for kp in result.keypoints.xy.cpu().numpy()]
```

Regarding the inclusion of keypoint IDs as a special attribute, while it may seem redundant, it could indeed simplify the process for users with complex use cases. We appreciate your suggestion and will consider it for future updates to enhance usability.

If you encounter any further issues or have additional suggestions, please let us know. Your input is invaluable in helping us improve! 😊

**Comment by glenn-jocher** [comment]
Current implementation (color by keypoint) is a compromise for best results in the most common use cases, which tends to be human pose, but also it's important to consider the official repo as a starting point for your own customization journey (i.e. color by class if you prefer), as it's not possible for us to create solutions for every use case.

**Comment by d-kleine** [comment]
> Current implementation (color by keypoint) is a compromise for best results in the most common use cases, which tends to be human pose, but also it's important to consider the official repo as a starting point for your own customization journey (i.e. color by class if you prefer), as it's not possible for us to create solutions for every use case.

Sorry, but do you even read issues carefully, or do you just ran them through Perplexity, generating an automated response?! Your answers usually contain points that are not even linked to the original issue.

This issue is NOT about coloring the keypoints (that is a different, the bug report #14389). This one here is a feature request is about the **keypoints ID labels**. Currently, there is keypoints ID labels in the output of `model.train()`. It would be great to have the ability to display the keypoint ID labels (and/or, if possible keypoint name labels) along with the keypoints, make with `model.predict`.

**Comment by Y-T-G** [comment]
> > Yeah, you have index into the particular object first. But still, I don't see why you need to add it as a special attribute, when it's a simple `enumerate(keypoints)` away.
> > ```python
> > kp_with_id = [[(i, x, y) for i, (x,y) in enumerate(kp)] for kp in result.keypoints.xy.cpu().numpy()]
> > ```
> 
> This doesn't work either when having classes with a different number of classes. I have tested with with my case: https://github.com/d-kleine/barrel_detection/blob/a1a4480ac2cdc01e49876b9dca49b70a70ad59af/2_inference.ipynb
> 
> I two classes classes: turret (4 keypoints) and hull (5 keypoints) For the "turret" class it create a new keypoint at `(0.0, 0.0)`:
> 
> ```
> [[(0, 153.64124, 337.01785),
>   (1, 450.85718, 328.95575),
>   (2, 414.2162, 355.46518),
>   (3, 218.0868, 364.19257),
>   (4, 0.0, 0.0)],
>  [(0, 192.04353, 330.56403),
>   (1, 488.22897, 354.98108),
>   (2, 544.5025, 434.77182),
>   (3, 166.18379, 449.05716),
>   (4, 595.5581, 375.10532)]]
> ```

I don't see how that's not working. The output is correct. You can't have different number of keypoints for different classes. All classes have the same number of keypoints and hence output has 5 keypoints for all classes. It will show that even if an internal keypoint id feature was implemented.

> 
> > The labels feature makes sense, but the ids is redundant.
> 
> It makes things so easier, so why not?

You want a feature that you can just use `enumerate(keypoints)` to obtain like I showed. It would have made sense if it required multiple lines of code to obtain the keypoint ids. But it doesn't.

**Comment by Y-T-G** [comment]
Usually the way you process keypoints is you would be defining an enum for each ID so that you can use them index them with the names easily. That would be a better feature suggestion alongside keypoint labels. And if you have an enum, you don't need the keypoint ids.

```python
class Tank(Enum):
    LEFT_WHEEL = 1
    RIGHT_WHEEL = 2
    ....

for box, kps in zip(result.boxes, result.keypoints):
    if kps[Tank.LEFT_WHEEL][0] > x:
        # Do something
```

**Comment by d-kleine** [comment]
> I don't see how that's not working. The output is correct. You can't have different number of keypoints for different classes. All classes have the same number of keypoints and hence output has 5 keypoints for all classes. It will show that extra empty keypoint even if an internal keypoint id feature was implemented. The correct way to handle this is to use the x, y, visibility format. So you can find out which keypoints are missing.

Yeah, sorry, I was trying to say that your solution works but only when removing the `[0, 0]` keypoints. @glenn-jocher has posted code (seems to have been removed) based your solution code snippet that works fine for my use case:
```python
# Extracting kpt IDs with kpt coordinates
kp_with_id = [[(i, x, y) for i, (x, y) in enumerate(kp) if (x, y) != (0.0, 0.0)] for kp in keypoints_data]
display(kp_with_id)
```
https://github.com/d-kleine/barrel_detection/blob/ffdcf813084e36621dffd8df63e36e3bb049b47f/2_inference.ipynb

So, both of you, thanks for that!

> You want a feature that you can just use `enumerate(keypoints)` to obtain like I showed. It would have made sense if it required multiple lines of code to obtain the keypoint ids. But it doesn't.

For usability and a positive user experience - it's easier for users to extract the keypoint IDs with an in-build functionality (attribute) than writing (some lines of) code by yourself.

> Usually the way you process keypoints is you would be defining an enum for each ID so that you can use them index them with the names easily. That would be a better feature suggestion alongside keypoint labels. And if you have an enum, you don't need the keypoint ids.

That was actually my idea, see #14405. But with the current YOLOv8 format, there are no keypoint name labels provided, you would need to provide them manually. Actually it would be better to have both keypoints ID and keypoint name provided for each keypoint as a dict., like
```python
data = {
    "person": {
        0: "nose",
        1: "left_eye",
        2: "right_eye",
        3: "left_ear",
        4: "right_ear"
    },
    "car": {
        0: "front_left_wheel",
        1: "front_right_wheel",
        2: "rear_left_wheel",
        3: "rear_right_wheel",
        4: "roof"
    }
}
```
So you can do:

```python
# Extract keypoint ID 3 from the "car" dictionary
car_rear_right_wheel = data["car"][3]

# Print the result
print(car_rear_right_wheel)
```

**Comment by glenn-jocher** [comment]
Thank you for your feedback. I understand your point about usability and the convenience of having built-in functionality for extracting keypoint IDs and labels. While the current implementation requires some manual steps, your suggestion to include keypoint IDs and labels directly in the results object is noted and could enhance user experience. For now, using the `x, y, visibility` format is the recommended approach to handle keypoints effectively. If you have further questions or need additional assistance, feel free to ask.
